Kivy: use the same password for all wallets

ecdsa opened the pull request:
When a wallet password is updated, check the password for all
wallets in directory. If that check passes, update password for
all wallets. If it fails, update only the current wallet.

ecdsa commented:
note: once this is enabled, we should let users reset their PIN with their password

SomberNight commented on electrum/gui/kivy/main_window.py:
```diff
@@ -649,6 +649,10 @@ def load_wallet_by_name(self, path):
             return
+        if self.password and self.electrum_config.get('use_single_password'):
+            storage = WalletStorage(path)
+            storage.check_password(self.password)
```
what if wallet is only keystore-encrypted?

ecdsa replied:
This case is handled in `check_password_for_directory`. `self._use_single_password` is set only if the password was checked against all wallets and keystores. Here check_password is called for its side effect, to decrypt the storage. I will add a comment to clarify.

SomberNight commented on electrum/gui/kivy/main_window.py:
```diff
@@ -1213,11 +1218,60 @@ def check_pin_code(self, pin):
+            path = os.path.join(dirname, filename)
+            basename = os.path.basename(path)
+            storage = WalletStorage(path)
+            if not storage.is_encrypted():
```
what is wallet is already keystore-encrypted (but not storage-encrypted)?

ecdsa replied:
this is now handled.

SomberNight commented on electrum/gui/kivy/main_window.py:
```diff
@@ -1213,11 +1218,60 @@ def check_pin_code(self, pin):
+            basename = os.path.basename(path)
+            storage = WalletStorage(path)
+            if not storage.is_encrypted():
+                # let's encrypt it
+                db = WalletDB(storage.read(), manual_upgrades=False)
+                wallet = Wallet(db, storage, config=self.electrum_config)
```
maybe these lines too should be guarded with `if update`?

ecdsa replied:
ack

SomberNight commented on electrum/gui/kivy/main_window.py:
```diff
@@ -1213,11 +1218,60 @@ def check_pin_code(self, pin):
+                wallet = Wallet(db, storage, config=self.electrum_config)
+                success.append(basename)
+                if update:
+                    wallet.update_password(None, new_password)
```
I guess `InvalidPassword` is not handled as you assume it cannot be raised due to `if not storage.is_encrypted()` - but again, wallet might be keystore-encrypted

ecdsa replied:
same as above

SomberNight commented on electrum/gui/kivy/main_window.py:
```diff
@@ -1213,11 +1218,60 @@ def check_pin_code(self, pin):
         if pin != self.electrum_config.get('pin_code'):
             raise InvalidPassword
 
+    def check_password_for_directory(self, old_password, new_password, update=False):
```
Could you either add a docstring or change the API?
E.g. 
```
"""Checks password against all wallets and returns (list of matches, list of mismatches).
If update is True, it also changes the wallet passwords to new_password.
"""
```
Alternatively, the API might be clearer if "update" and "new_password" are merged together. Simplest would be to set `new_password` to `None` to correspond to `update==False`, however `None` is a valid value for `new_password` I guess, so we could introduce a sentinel e.g.
```
SENTINEL=object()
def check_password_for_directory(self, old_password, new_password=SENTINEL): ...
```

ecdsa replied:
I agree, it is better to change the API. `None` is not a valid password for Kivy, so I guess it is fine to use it as long as this method stays in the GUI

SomberNight commented on electrum/gui/kivy/main_window.py:
```diff
@@ -1213,11 +1218,60 @@ def check_pin_code(self, pin):
+            if not failed:
+                self.check_password_for_directory(old_password, new_password, update=True)
+                self.electrum_config.set_key('use_single_password', True)
+                msg = _(f"Password updated successfully for {len(success)} wallets")
```
this does not work, see https://github.com/spesmilo/electrum/blob/78513affe536ba7d8150dc2ee044e3258097317f/electrum/i18n.py#L33-L35

ecdsa commented:
note: if a user does not know the unique password, they can still create a new wallet, and this will result in a situation where the unique password is disabled. This could be annoying in case of an unauthorized third party accessing the phone.

ecdsa commented:
Rebased and squashed to have a single commit. Wallet creation is now disabled if the user does not know their password, so that password unification is irreversible.

An edge case remains: if the first wallet to be open is unencrypted and watching-only, no password is asked and `use_single_password` will be set to False
